do/csi/driver/controller.go

```diff
@@ -3,48 +3,347 @@ package driver
 import (
 	"context"
 	"errors"
+	"fmt"
+	"net/http"
+	"strconv"
+	"strings"
 
 	csi "github.com/container-storage-interface/spec/lib/go/csi/v0"
+	"github.com/digitalocean/godo"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
 )
 
-// CreateVolume creates a new volume
-func (d *Driver) CreateVolume(context.Context, *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
-	return nil, errors.New("not implemented")
+const (
+	_  = iota
+	KB = 1 << (10 * iota)
+	MB
+	GB
+	TB
+)
+
+const (
+	defaultVolumeSizeInGB = 16 * GB
+
+	createdByDO = "Created by DigitalOcean CSI driver"
+)
+
+// CreateVolume creates a new volume from the given request. The function is
+// idempotent.
+func (d *Driver) CreateVolume(ctx context.Context, req *csi.CreateVolumeRequest) (*csi.CreateVolumeResponse, error) {
+	if req.Name == "" {
+		return nil, status.Error(codes.InvalidArgument, "CreateVolume Name must be provided")
+	}
+
+	if req.VolumeCapabilities == nil || len(req.VolumeCapabilities) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "CreateVolume Volume capabilities must be provided")
+	}
+
+	size, err := extractStorage(req.CapacityRange)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	volumeName := req.Name
+
+	// get volume first, if it's created do no thing
+	volumes, _, err := d.doClient.Storage.ListVolumes(ctx, &godo.ListVolumeParams{
```

Comment:
Something we can do later, but if you don't paginate you'll only get the first 25 volumes.

Reply:
oh nvm, seems like we're doing a filter on volume name so don't think it matters anyways